src/tls_crypt_v2.ml

```diff
@@ -0,0 +1,329 @@
+let error_msgf fmt = Fmt.kstr (fun msg -> Error (`Msg msg)) fmt
+let guard ~msg test = if test () then Ok () else Error (`Msg msg)
+
+module List = struct
+  include List
+
+  let chop lst =
+    match List.rev lst with
+    | [] -> invalid_arg "List.chop"
+    | _ :: lst -> List.rev lst
+
+  let trim lst =
+    List.fold_left (fun acc -> function "" -> acc | str -> str :: acc) [] lst
+    |> List.rev
+end
+
+module String = struct
+  include String
+
+  let split_at n str =
+    let rec go acc str off len =
+      if len = 0 then List.rev acc
+      else
+        let len' = min n len in
+        let str' = String.sub str off len' in
+        go (str' :: acc) str (off + len') (len - len')
+    in
+    go [] str 0 (String.length str)
+end
+
+let _TLS_CRYPT_V2_CLIENT_KEY_LEN = 2048 / 8
+let _TLS_CRYPT_V2_MAX_WKC_LEN = 1024
+let _TLS_CRYPT_V2_TAG_SIZE = 256 / 8
+let _TLS_CRYPT_V2_MAX_METADATA_LEN = _TLS_CRYPT_V2_MAX_WKC_LEN - (_TLS_CRYPT_V2_CLIENT_KEY_LEN + _TLS_CRYPT_V2_TAG_SIZE + 2)
+
+let pem_of_lines ~name seq =
+  match List.of_seq seq with
+  | [ _ ] -> error_msgf "Invalid %s" name
+  | header :: b64_and_footer when header = "-----BEGIN " ^ name ^ "-----" ->
+      let b64_and_footer = List.trim b64_and_footer in
+      let footer = List.hd (List.rev b64_and_footer) in
+      if footer <> "-----END " ^ name ^ "-----" then
+        error_msgf "Invalid %s" name
+      else
+        let b64 = String.concat "" (List.chop b64_and_footer) in
+        Base64.decode b64
+  | _ -> error_msgf "Invalid %s" name
+
+module Key : sig
+  type t
+
+  val of_cstruct : Cstruct.t -> (t, [> `Msg of string ]) result
+  val unsafe_to_cstruct : t -> Cstruct.t
+  val to_string : t -> string
+  val cipher_key : t -> Mirage_crypto.Cipher_block.AES.CTR.key
+  val hmac : t -> Cstruct.t
+  val equal : t -> t -> bool
+  val pp_hum : t Fmt.t
+  val generate : ?g:Mirage_crypto_rng.g -> unit -> t
+  val to_base64 : t -> string
+end = struct
+  type t = Cstruct.t (* cipher (64 bytes) + hmac (64 bytes) *)
+
+  let len_of_secret_aes_ctr_key =
+    Array.to_list Mirage_crypto.Cipher_block.AES.CTR.key_sizes
+    |> List.filter ((>) 128) (* ∀x. 128 > x <=> ∀x. x <= 128 *)
+    |> List.fold_left max 0
+
+  (* NOTE: it's a bit paranoid but we ensure that [mirage-crypto] exposes values
+     which fit with an OpenVPN key (128 bytes). *)
+  let () = assert (len_of_secret_aes_ctr_key > 0)
```

Comment:
I would go with `((>=) 64)` as each key (128 bytes) is a pair of cipher key material (up to 64 bytes) and hmac key material (up to 64 bytes).